lib/tls-gnutls.c

```diff
@@ -319,6 +382,268 @@ psk_srv(gnutls_session_t session, const char *username, gnutls_datum_t *key)
+                                         &keydata,
+                                         GNUTLS_X509_FMT_PEM,
+                                         pin,
+                                         0);
+    if (ret < 0)
+      goto out;
+  }
+
+  *pcert = p;
+  *pcert_length = cert_size;
+  *pkey = k;
+
+out:
+  explicit_bzero(keydata.data, keydata.size);
+  gnutls_free(keydata.data);
+
+  if (ret != 0) {
+    if (p != NULL) {
+      clear_plist(p, cert_size);
+      free(p);
+    }
+    if (k != NULL) {
+      gnutls_privkey_deinit(k);
+    }
+    return -1;
+  }
+  return 0;
+}
+
+static int
+cert_srv(gnutls_session_t session, const gnutls_datum_t *req_ca_dn, int nreqs, const gnutls_pk_algorithm_t *pk_algos, int pk_algos_length, gnutls_pcert_st **pcert, unsigned int *pcert_length, gnutls_privkey_t *pkey)
+{
+  const tls_handshake_t *hs = gnutls_session_get_ptr(session);
+  char *cert_uri = NULL;
+  char *key_uri = NULL;
+  char *pin = NULL;
+  int ret = 0;
+  /* DNS names (only type supported) may be at most 256 bytes long */
+  char *servername;
+  size_t namelen = 256;
+  unsigned int nametype;
+
+  servername = gnutls_malloc(namelen);
+  if (servername == NULL)
+    return GNUTLS_E_MEMORY_ERROR;
+
+  for(int i = 0; ; i++) {
+    ret = gnutls_server_name_get(session, servername, &namelen, &nametype, i);
+    if (ret == GNUTLS_E_REQUESTED_DATA_NOT_AVAILABLE) {
+      /* We hit no NAME_DNS before the end, so no servername we can use. But that might be fine. */
+      free(servername);
+      servername = NULL;
+      break;
+    }
+    if (ret == GNUTLS_E_SHORT_MEMORY_BUFFER)
+      continue;  // This must not be DNS, or an invalid DNS name. Ignore
+    if (ret != GNUTLS_E_SUCCESS)
+      goto out;
+    if (nametype != GNUTLS_NAME_DNS)
+      continue;
+    /* This was a correctly retrieved DNS name. Per RFC6066, section 3, that means there
+        are no more DNS names in this request. Send this forward. */
+    break;
+  }
+
+  ret = hs->srv.cert.getcert(hs->srv.misc, servername, &cert_uri, &key_uri, &pin);
+  if (ret < 0)
+    goto out;
+
+  ret = load_certificate(session, pcert, pcert_length, pkey, cert_uri, key_uri, pin);
+
+out:
+  if (servername != NULL) {
+    free(servername);
+  }
+  if (cert_uri != NULL) {
+    explicit_bzero(cert_uri, strlen(cert_uri));
+    free(cert_uri);
+  }
+  if (key_uri != NULL) {
+    explicit_bzero(key_uri, strlen(key_uri));
+    free(key_uri);
+  }
+  if (pin != NULL) {
+    explicit_bzero(pin, strlen(pin));
+    free(pin);
+  }
+  return ret;
+}
+
+static int
+cert_clt(gnutls_session_t session, const gnutls_datum_t *req_ca_dn, int nreqs, const gnutls_pk_algorithm_t *pk_algos, int pk_algos_length, gnutls_pcert_st **pcert, unsigned int *pcert_length, gnutls_privkey_t *pkey)
+{
+  const tls_handshake_t *hs = gnutls_session_get_ptr(session);
+  char *cert_uri = NULL;
+  char *key_uri = NULL;
+  char *pin = NULL;
+  int ret = 0;
+  char **req_ca_dn_strs = gnutls_malloc(sizeof(char *) * (nreqs + 1));
```

Comment:
Please move the `malloc()` call out of the variable declaration block.  (Put it next to the `NULL` check.)